expression/builtin_time.go

```diff
@@ -3280,1581 +3388,616 @@ func (du *baseDateArithmetical) vecGetIntervalFromReal(b *baseBuiltinFunc, input
 	return nil
 }
 
-type addDateFunctionClass struct {
+type funcTimeOpForDateAddSub func(da *baseDateArithmetical, ctx sessionctx.Context, date types.Time, interval string, unit string, resultFsp int) (types.Time, bool, error)
+
+func addTime(da *baseDateArithmetical, ctx sessionctx.Context, date types.Time, interval string, unit string, resultFsp int) (types.Time, bool, error) {
+	return da.add(ctx, date, interval, unit, resultFsp)
+}
+
+func subTime(da *baseDateArithmetical, ctx sessionctx.Context, date types.Time, interval string, unit string, resultFsp int) (types.Time, bool, error) {
+	return da.sub(ctx, date, interval, unit, resultFsp)
+}
+
+type funcDurationOpForDateAddSub func(da *baseDateArithmetical, ctx sessionctx.Context, d types.Duration, interval string, unit string, resultFsp int) (types.Duration, bool, error)
+
+func addDuration(da *baseDateArithmetical, ctx sessionctx.Context, d types.Duration, interval string, unit string, resultFsp int) (types.Duration, bool, error) {
+	return da.addDuration(ctx, d, interval, unit, resultFsp)
+}
+
+func subDuration(da *baseDateArithmetical, ctx sessionctx.Context, d types.Duration, interval string, unit string, resultFsp int) (types.Duration, bool, error) {
+	return da.subDuration(ctx, d, interval, unit, resultFsp)
+}
+
+type funcSetPbCodeOp func(b builtinFunc, add tipb.ScalarFuncSig, sub tipb.ScalarFuncSig)
+
+func setAdd(b builtinFunc, add tipb.ScalarFuncSig, sub tipb.ScalarFuncSig) {
+	b.setPbCode(add)
+}
+
+func setSub(b builtinFunc, add tipb.ScalarFuncSig, sub tipb.ScalarFuncSig) {
+	b.setPbCode(sub)
+}
+
+type funcGetDateForDateAddSub func(da *baseDateArithmetical, ctx sessionctx.Context, args []Expression, row chunk.Row, unit string) (types.Time, bool, error)
+
+func getDateFromString(da *baseDateArithmetical, ctx sessionctx.Context, args []Expression, row chunk.Row, unit string) (types.Time, bool, error) {
+	return da.getDateFromString(ctx, args, row, unit)
+}
```

Comment:
Are there other implementations of all these getDate[time]From{String,Int,Real,Decimal} etc. in other places of the repository that can be combined with these new ones?

Reply:
I did have the same concern as yours and did some searching. Didn't see other usage of `baseDateArithmetical` other than `date_add/sub`.

Reply:
Thanks for looking!